test/test_common/logging.h

```diff
@@ -61,6 +61,32 @@ class LogRecordingSink : public Logger::SinkDelegate {
   std::vector<std::string> messages_;
 };
 
+typedef std::vector<std::pair<std::string, std::string>> ExpectedLogSequence;
+
+// Validates that when the stmt is executed, a sequence of emmitted log messages matches the
+// sequence of expected logs/log level pair (ExpectedLogSequence). Note that both sequences must
+// respect the same order.
+#define EXPECT_LOG_SEQ(expected_log_sequence, stmt)                                                \
+  do {                                                                                             \
+    LogLevelSetter save_levels(spdlog::level::trace);                                              \
+    LogRecordingSink log_recorder(Logger::Registry::getSink());                                    \
+    stmt;                                                                                          \
+    ASSERT_EQ(expected_log_sequence.size(), log_recorder.messages().size());                       \
+    auto expected_it = expected_log_sequence.begin();                                              \
+    auto actual_it = log_recorder.messages().begin();                                              \
+    for (;                                                                                         \
+         expected_it != expected_log_sequence.end() && actual_it != log_recorder.messages().end(); \
+         expected_it++, actual_it++) {                                                             \
+      /* Parse "[2018-04-02 19:06:08.629][15][warn][admin] source/file.cc:691] message ..." */     \
+      std::vector<absl::string_view> pieces = absl::StrSplit(*actual_it, "][");                    \
+      const auto expected_log_level = expected_it->first;                                          \
+      const auto actual_log_level = std::string(pieces[2]);                                        \
+      const auto actual_log_message = absl::string_view(pieces[3]).find(expected_it->second);      \
+      EXPECT_EQ(expected_log_level, actual_log_level);                                             \
+      EXPECT_NE(actual_log_message, absl::string_view::npos);                                      \
```

Comment:
Actually, one other thing: I think when this fails, the error messages won't be very helpful.  See the EXPECT_TRUE below in EXPECT_LOG_CONTAINS for how I go out of the way to give an immediately actionable error message, so that typos or changes can be resolved based on the test results log, without having to debug.

Thinking about this further, if you have a multi-line comparison it probably would be better to join everything as strings and just do a string comparison.  Consider the case where a new message gets added.  You want to see what that message was in the test failure log, but here you won't get that far; you'll assert on the size-difference.

Reply:
@jmarantz That's a great suggestion. I will add a todo to for the prettification of the error message. I'm working on some other issues now, but I can come back for this one later.